src/GitHub.Api/Application/ApplicationManagerBase.cs

```diff
@@ -179,29 +179,24 @@ private void InitializeEnvironment(NPath gitExecutablePath)
 
             if (Environment.IsWindows)
             {
-                GitClient
+                var task = GitClient
                     .GetConfig("credential.helper", GitConfigSource.Global)
                     .Then((b, credentialHelper) => {
-                        if (!string.IsNullOrEmpty(credentialHelper))
+                        if (string.IsNullOrEmpty(credentialHelper))
                         {
-                            Logger.Trace("Windows CredentialHelper: {0}", credentialHelper);
-                            afterGitSetup.Start();
+                            Logger.Warning("No Windows CredentialHelper found: Setting to wincred");
+                            throw new ArgumentNullException(nameof(credentialHelper));
                         }
-                        else
-                        {
-                            Logger.Warning("No Windows CredentialHeloper found: Setting to wincred");
+                    });
+                // if there's no credential helper, set it before restarting the repository
+                task.Then(GitClient.SetConfig("credential.helper", "wincred", GitConfigSource.Global), TaskRunOptions.OnFailure)
+                    .Then(afterGitSetup, taskIsTopOfChain: true);
```

Comment:
What effect does `taskIsTopOfChain` achieve?

Reply:
Let's say we have chains of tasks like this:

chain1: A->B
chain2: M->N
chain3: Y->Z

A is the top of the chain1. B depends on A,. Some method builds this chain and returns B.
M is the top of chain2. N depends on M. Some method builds this chain and returns N.
Y is the top of chain3. Z depends on Y. We build this chain and have a reference to Y.

Y is a chain that we want to always call when A->B and M->N finish. We want:

A->B->Y->Z
M->N->Y->Z

We call `B.Then(Y)`. The default behaviour of `Then` is to go up the chain of the task we're appending to grab the top of the chain. In this case, we're passing it Y, which is the top, so it appends Y to B, creating A->B->Y->Z.

We call `N.Then(Y)`. `Then` goes up the chain of Y to get the top of it, which is **now A**, and **appends A to N**. We now have M->N->A->B->Y->Z. Oops.

`taskIsTopOfChain` tells `Then` not to go up the chain to get the root, and instead just append the one we're sending. This way, we can call `N.Then(Y, taskIsTopOfChain: true)` and it will *not* go up the Y tree and will just append Y to N, creating M->N->Y->Z

Reply:
BTW, I've since found a better way to handle calling different tasks depending on the return of the previous task, by hooking up the `OnEnd` event and appending (calling Then) there depending on the result. Tasks won't be scheduled until after `OnEnd` is called, and `OnEnd` is guaranteed to be called for the current task and it provides the task itself as an argument, so it's a much nicer way of switching tasks.

I might take out the OnFailure task codepath completely since between Finally and OnEnd there are already decent ways of handling this situation.

Reply:
So after going through your comments let me interpret this hunk of code..

https://github.com/github-for-unity/Unity/blob/e7415510a0f7e2085f43d3d26409d9419829bf14/src/GitHub.Api/Application/ApplicationManagerBase.cs#L174-L199

1. We setup `afterGitSetup` with two tasks in it. 
**A -> B**
2. If we are on windows
  a. We setup two tasks that read a config and throw if it's now found. 
**C -> D**
  b. We create a task that sets the config. **E** And schedule it to run on **D**'s failure with `afterGitSetup` to run after it.
**C -> D -> (fail: E -> A -> B)**
*question: you are using `taskIsTopofChain` here but that has no effect here*
  c. We also set `afterGitSetup` to run on **D**'s success. 
**C -> D -> (success: A -> B ; fail: E -> A -> B)**
*question: you are using `taskIsTopofChain` here but that is necessary here*
3. Start `afterGitSetup`

Reply:
Yeah, that's it. I'm using `taskIsTopOfChain` on both because it's very easy to reorder those instructions and break without realizing it. The intent is to append that task and its children *only** onto the chain. Whether there are more tasks above it or not is a detail I don't want to have to worry about later, basically.